samples/sobel/sobel_main.py

```diff
@@ -1,56 +1,231 @@
 
-   r = hcl.reduce_axis(0,3)
-   c = hcl.reduce_axis(0,3)
-   B = hcl.compute((height-2,width-2), 
-           lambda x,y: hcl.sum(A[x+r,y+c]*Gx[r,c], axis=[r,c]),
-           name="B", dtype=hcl.Float())
-   t = hcl.reduce_axis(0,3)
-   g = hcl.reduce_axis(0,3)
+    r = hcl.reduce_axis(0, 3)
+    c = hcl.reduce_axis(0, 3)
+    D = hcl.compute((height-2, width-2), 
+        lambda x, y: hcl.sum(B[x+r, y+c]*Gx[r, c], axis=[r, c], name = "sum1"), "D")
 
-   C = hcl.compute((height-2,width-2), 
-           lambda x,y: hcl.sum(A[x+t,y+g]*Gy[t,g], axis=[t,g]),
-           name="C", dtype=hcl.Float())
-   return hcl.compute((height-2,width-2), 
-              lambda x, y :hcl.sqrt(B[x,y]*B[x,y] + C[x,y]*C[x,y])/4328*255,
-              dtype=hcl.Float())
+    t = hcl.reduce_axis(0, 3)
+    g = hcl.reduce_axis(0, 3)
+    E = hcl.compute((height-2, width-2), 
+        lambda x, y: hcl.sum(B[x+t, y+g]*Gy[t, g], axis=[t, g], name = "sum2"), "E")
 
-s = hcl.create_schedule([A,Gx,Gy],sobel)
-f = hcl.build(s)
+    # constant factor to normalize the output
+    return hcl.compute((height-2,width-2), 
+        lambda x, y:hcl.sqrt(D[x][y]*D[x][y]+E[x][y]*E[x][y])*0.05891867, "Fimg")
 
-npA = np.array(img)
-npGx = np.array([[1,0,-1],[2,0,-2],[1,0,-1]])
-npGy = np.array([[1,2,1],[0,0,0],[-1,-2,-1]])
-hcl_A = hcl.asarray(npA)
+# create a schedule
+s = hcl.create_schedule([A, Gx, Gy], sobel)
+
+###############################################################################
+# Optimization
+# ============
+# HeteroCL provides different primitives to optimize the performance of the
+# program. In this example, since convolutions involve overlapped access to
+# the same memory location, we can make use of the idea of window- and line-
+# buffer to optimize for memory usage. The `reuse_at` primitive supports the
+# data reuse in the specified axis.
+# For further optimization, we apply the `partition` and `pipeline` primitives 
+# to maximize the memory bandwidth via loop transformation.
+
+LBX = s.reuse_at(sobel.B, s[sobel.D], sobel.D.axis[0], "LBX")
+LBY = s.reuse_at(sobel.B, s[sobel.E], sobel.E.axis[0], "LBY") 
+WBX = s.reuse_at(LBX, s[sobel.D], sobel.D.axis[1], "WBX")
+WBY = s.reuse_at(LBY, s[sobel.E], sobel.E.axis[1], "WBY")
+s.partition(LBX, dim=1)
+s.partition(LBY, dim=1)
+s.partition(WBX)
+s.partition(WBY)
+s.partition(Gx)
+s.partition(Gy)
+s[sobel.B].pipeline(sobel.B.axis[1])
+s[sobel.D].pipeline(sobel.D.axis[1])
+s[sobel.E].pipeline(sobel.E.axis[1])
+s[sobel.Fimg].pipeline(sobel.Fimg.axis[1])
+
+###############################################################################
+# Results
+# =======
+# We explicitely and numerically define the inputs to the Sobel operation and
+# perform the execution. To convert an input to a HeteroCL-compatible type, we
+# use `.asarray()` primitive to convert the input into HeteroCL array.
+#
+# OPTIONAL: You can specify the FPGA target to run under using the `target`
```

Comment:
This paragraph looks kind of confusing. Are you trying to say -- If the `target` is missing in `hcl.build`, then it will run CPU simulation? To use the HLS report feature, this target is required, otherwise HLS tool would not know which hardware it is synthesizing for.